examples_tests/49.NewAPITest/main.cpp

```diff
@@ -0,0 +1,330 @@
+		auto fs_unspec = device->createGPUShader(core::make_smart_refctd_ptr<asset::ICPUShader>(fs_source));
+
+		asset::ISpecializedShader::SInfo vsinfo(nullptr, nullptr, "main", asset::ISpecializedShader::ESS_VERTEX, "vs");
+		auto vs = device->createGPUSpecializedShader(vs_unspec.get(), vsinfo);
+		asset::ISpecializedShader::SInfo fsinfo(nullptr, nullptr, "main", asset::ISpecializedShader::ESS_FRAGMENT, "fs");
+		auto fs = device->createGPUSpecializedShader(fs_unspec.get(), fsinfo);
+
+		video::IGPUSpecializedShader* shaders[2]{ vs.get(), fs.get() };
+
+		asset::SVertexInputParams vtxinput;
+		vtxinput.attributes[0].binding = 0;
+		vtxinput.attributes[0].format = asset::EF_R32G32_SFLOAT;
+		vtxinput.attributes[0].relativeOffset = offsetof(SVertex, pos);
+
+		vtxinput.attributes[1].binding = 0;
+		vtxinput.attributes[1].format = asset::EF_R32G32B32_SFLOAT;
+		vtxinput.attributes[1].relativeOffset = offsetof(SVertex, color);
+
+		vtxinput.bindings[0].inputRate = asset::EVIR_PER_VERTEX;
+		vtxinput.bindings[0].stride = sizeof(SVertex);
+
+		vtxinput.enabledAttribFlags = 0b0011;
+		vtxinput.enabledBindingFlags = 0b0001;
+		
+		asset::SRasterizationParams raster;
+		raster.depthTestEnable = 0;
+		raster.depthWriteEnable = 0;
+		raster.faceCullingMode = asset::EFCM_NONE;
+		
+		asset::SPrimitiveAssemblyParams primitive;
+		primitive.primitiveType = asset::EPT_TRIANGLE_LIST;
+
+		asset::SBlendParams blend;
+
+		rpindependent_pipeline = device->createGPURenderpassIndependentPipeline(nullptr, core::smart_refctd_ptr(layout), shaders, shaders+2, vtxinput, blend, primitive, raster);
+		assert(rpindependent_pipeline);
+	}
+
+	core::smart_refctd_ptr<video::IGPUGraphicsPipeline> pipeline;
+	{
+		video::IGPUGraphicsPipeline::SCreationParams gp_params;
+		gp_params.rasterizationSamplesHint = asset::IImage::ESCF_1_BIT;
+		gp_params.renderpass = renderpass;
+		gp_params.renderpassIndependent = rpindependent_pipeline;
+		gp_params.subpassIx = 0u;
+
+		pipeline = device->createGPUGraphicsPipeline(nullptr, std::move(gp_params));
+	}
+
+	core::smart_refctd_ptr<video::IGPUBuffer> buffer;
+	{
+		const SVertex vertices[3]{
+			{{0.0f, -0.5f}, {1.0f, 0.0f, 0.0f}},
+			{{0.5f, 0.5f}, {0.0f, 1.0f, 0.0f}},
+			{{-0.5f, 0.5f}, {0.0f, 0.0f, 1.0f}}
+		};
+
+		video::IDriverMemoryBacked::SDriverMemoryRequirements mreq;
+		
+		buffer = device->createDeviceLocalGPUBufferOnDedMem(sizeof(vertices));
+		assert(buffer);
+
+		core::smart_refctd_ptr<video::IGPUCommandBuffer> cb;
+		device->createCommandBuffers(cmdpool.get(), video::IGPUCommandBuffer::EL_PRIMARY, 1u, &cb);
+		assert(cb);
+
+		cb->begin(video::IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+		cb->updateBuffer(buffer.get(), 0u, sizeof(vertices), vertices);
+		cb->end();
+		
+		video::IGPUQueue::SSubmitInfo info;
+		auto* cb_ = dynamic_cast<video::IGPUPrimaryCommandBuffer*>(cb.get());
+		info.commandBufferCount = 1u;
+		info.commandBuffers = &cb_;
```

Comment:
an interesting question, does your queue create smart refcounted pointers to the commandbuffers which have been submitted?

If so how will we ensure that they are not dropped until the GPU is actually done?

Reply:
we wont, why would we? cmdbuf doesnt represent any GL object and even if it did, GL decides when resources are actually deleted itself

> does your queue create smart refcounted pointers to the commandbuffers which have been submitted ?

yes

Reply:
> we wont, why would we? cmdbuf doesnt represent any GL object and even if it did, GL decides when resources are actually deleted itself

Uniformity between the APIs, when the user will switch to Vulkan, suddenly the app will crash because the're no mechanism for the deferred dropping/deletion of Vulkan commandbuffer objects (and others).

Reply:
ok the commands in the commandbuffer (in the future pool, cause I want the commands allocated from the pool and have the ordering implicit from a linked list, not a vector per commandbuffer) hold smartpointers, and we know how to create a mechanism for holding commandbuffers alive as long as necessary in userspace (deferred event handler)